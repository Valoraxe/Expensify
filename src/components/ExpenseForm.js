import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount:  props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      toEdit: props.toEdit
    }
  }
  descriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }
  noteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }
  amountChange = (e) => {
    const amount = e.target.value
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  changeDate = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({error: "Please set description & amount"}))
      this.state.errorOnSubmit = true
    } else {
      this.setState(() => ({error: ""}))
      this.state.errorOnSubmit = false
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form-error">{this.state.error}</p>}
        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.descriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.amountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.changeDate}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add a note"
          className="textarea"
          value={this.state.note}
          onChange={this.noteChange}
        >
        </textarea>
        <div>
          {this.state.toEdit
            ? <button className="button">Save Changes</button>
            : <button className="button">Add Expense</button>
          }
        </div>
      </form>
    )
  }
}
