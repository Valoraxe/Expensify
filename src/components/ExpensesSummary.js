import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/ExpensesTotal'

export const ExpensesSummary = ({ visibleExpenseCount, invisibleExpenseCount, expensesTotal }) => {
  const expenseWord = visibleExpenseCount === 1 ? "expense" : "expenses"
  const invisibleWord = invisibleExpenseCount === 1 ? "expense" : "expenses"
  const areOrIs = invisibleExpenseCount === 1 ? "is" : "are"
  const formatTotal = numeral(expensesTotal / 100).format("0,0.00")
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header_title">
          Viewing <span>{visibleExpenseCount}</span> {expenseWord} totalling <span>Â£{formatTotal}</span>
        </h1>
        {invisibleExpenseCount === 0
          ? <h2 className="page-header_title">There are no hidden expenses</h2>
          : <h2 className="page-header_title">
              Currently <span>{invisibleExpenseCount}</span> {invisibleWord} {areOrIs} hidden due to the current filter
            </h2>
        }
        <div className="page-header_actions">
          <Link to="/create" className="button">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  const invisibleExpenses = state.expenses.length - visibleExpenses.length

  return {
    visibleExpenseCount: visibleExpenses.length,
    invisibleExpenseCount: invisibleExpenses,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
