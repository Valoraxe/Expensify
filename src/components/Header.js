import React from 'react'
import { Link } from 'react-router-dom'

const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header_content">
        <Link className="header_title" to="/" exact={true}>
          <h1>Expensify</h1>
        </Link>
        <button className="button button-link">Logout</button>
      </div>
    </div>
  </header>
)

export default Header
