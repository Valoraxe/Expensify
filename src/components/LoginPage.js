import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from '../actions/authentication'

export const LoginPage = ({ startLogin }) => (
  <div className="box-layout">
    <div className="box-layout_box">
      <h1 className="box-layout_title">Expensify</h1>
      <p>How much did you spend this month?</p>
      <button onClick={startLogin} className="button">Login with Google</button>
    </div>
  </div>
)

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
})

export default connect(undefined, mapDispatchToProps)(LoginPage)
