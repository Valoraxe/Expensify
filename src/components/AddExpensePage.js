import React from 'react'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push("/")
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header_title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            toEdit={this.props.toEdit}
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  toEdit: false
})

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage)
