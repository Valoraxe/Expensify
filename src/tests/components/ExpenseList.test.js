import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import ExpneseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'


test("render expense list", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})

test("render no expense list", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})
