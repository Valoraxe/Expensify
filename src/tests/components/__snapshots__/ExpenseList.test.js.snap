// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render expense list 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-mobile"
    >
      Expenses
    </div>
    <div
      className="show-desktop"
    >
      Expense
    </div>
    <div
      className="show-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={300}
      createdAt={0}
      description="Buddy"
      id="1"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={600}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={2000}
      createdAt={345600000}
      description="Credits"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;

exports[`render no expense list 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-mobile"
    >
      Expenses
    </div>
    <div
      className="show-desktop"
    >
      Expense
    </div>
    <div
      className="show-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item_message"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;
